File: leetcode/python/hard/1964_find_longest_valid_obstacle_course_at_each_position.py
import bisect


class Solution:
    def longestObstacleCourseAtEachPosition(self, obstacles: list[int]) -> list[int]:
        res = []
        seq = []
        for height in obstacles:
            idx = bisect.bisect_right(seq, height)
            if idx == len(seq):
                seq.append(height)
            else:
                seq[idx] = height
            res.append(idx + 1)
        return res

File: leetcode/rust/src/hard/1964_find_longest_valid_obstacle_course_at_each_position.rs
impl Solution {
    pub fn longest_obstacle_course_at_each_position(obstacles: Vec<i32>) -> Vec<i32> {
        let mut res = vec![];
        let mut seq = vec![];
        for o in obstacles {
            let i = seq.partition_point(|&x| x <= o);
            if i == seq.len() {
                seq.push(o);
            } else {
                seq[i] = o;
            }
            res.push((i + 1) as i32);
        }
        res
            
    }
}